### ISSUE TITLE
Ebean-elasticsearch paging query problem

### ISSUE THREAD

**Issue by UlyssesChou** [ebean-orm/ebean#2432]
Ebean-elasticsearch defaults to paging from+size.

However, in the case of deep paging, the efficiency of this method is very low. For example, from=5000,size=10, ES needs to match and sort on each fragment and get 5000*10 valid data, and then return the last 10 data in the result set. In addition to efficiency problems, there is also an unsolvable problem that ES currently supports the maximum skip value of max_result_window, which defaults to 10000, that is, when from+size &gt; When max_result_window, es returns an error.

So when I was in use Ebean, directly use the text () through the es query, is unable to satisfy my business scenario, because after more than 10000, access to the pager maximum page number is not correct, as far as I know the solution, when paging to hundreds of pages, es can't return data, at this time in order to restore the normal use, We can take an emergency dodge by setting max_result_window to 50000, but this solution is only a temporary fix.

So how do I properly page through Ebean using scroll, or something else I'm doing wrong, or is there a better solution

Looking forward to your reply

**Comment by rbygrave** [comment]
If you use an ebean streaming query like findEach() ... then that translates into an ElasticSearch scrolling query

findEach:
https://github.com/ebean-orm/ebean-elasticsearch/blob/master/src/test/java/integration/QueryEachTest.java#L30

scrolling query:
https://github.com/ebean-orm/ebean-elasticsearch/blob/master/src/main/java/io/ebeanservice/elastic/query/EQueryEach.java#L66

Is there a reason to not use `findEach()` for this case?

**Comment by ulyChou** [comment]
How do I get the total count of items hit by a query through the Ebean

**Comment by rbygrave** [comment]
Are you using `findEach()` for this case?

**Comment by ulyChou** [comment]
I want to use eBeans for scrolling paging

**Comment by ulyChou** [comment]
When I use from and size for deep paging,When I use `setFirstRow(Page * pageSize).setMaxRows(pageSize)`, the maximum pagedList.totalCount =10000, which does not fulfill the paging requirement correctly

**Comment by rbygrave** [comment]
> deep paging

Deep paging is pretty terrible for all the databases I know including ElasticSearch. DO NOT DO THIS.

> When I use setFirstRow(Page * pageSize).setMaxRows(pageSize)

DO NOT DO THIS when firstRow is very large - this is not a good thing to do for any database I know. Do not do this.


That is, "Paging" is good when the use case is mostly that users view / page the "first few pages" - when firstRow is not large.  "Paging" degrades when you go to deep paging (firstRow gets large) and we should very much avoid "Paging" for this case.  

"Scrolling" queries are good for the case where the query is expecting many many results. This is what ElasticSearch scrolling queries are for. We use Ebean `findEach()` / `findIterate()` / `findStream()` for this type of query and under the hood that is using an ElasticSearch scrolling query.

So, use `findEach()` - do **_NOT_** use `setFirstRow().findPagedList()` for this use case (when firstRow gets large).

So asking again, are you using `findEach()`?  If not why not?

**Comment by ulyChou** [comment]
Thank you very much for your reply :+1: 

Yes, with your guidance, I've now switched to findEach, but how do I get an effect similar to mysql's findCount,I need to assemble pagination myself.

**Comment by rbygrave** [comment]
I think you need to explain more about what you are trying to achieve.

That is, when we are doing a large query that is going to iterate through say a million results ... we pretty much should not care what the total count is. We should not care if we are processing 1,000,000 results or 1,000,001 results.  If the use case is that that we desire a total count as some estimate of the number of results we are going to process before we do the processing we could do a `findCount()` first but I'd suggest that isn't a common need or requirement.

So is this use case something that means we need a `findCount()` returning the expected number of hits (which could reasonably be a big number) and then use a `findEach()` query to process all those results. 

Can you outline better what the use case is that needs the _total count_ before it then processes all the results?  What is the purpose of the estimated _total count_ is this case?

**Comment by ulyChou** [comment]
Yes, my requirement is very simple, I want to use Scroll to achieve paging effect, I understand the problem you mentioned about the large amount of data, but I want to use scrollId and findCount to query the number of hits, so that I can achieve a correct pager similar to deep paging. Of course, the user doesn't have to span the pages too much, but I need to know how many findCount hits I have, and I can give the user a number cue and an option to go to the next page

**Comment by rbygrave** [comment]
Your requirement is not clear.  Things you say above suggest you have a user looking at UI / user interface / web page that would only contain a page of data and a total count?  This is not the use case for ElasticSearch scrolling query or ebean `findEach()`.

Can you clarify what the use case is. Do you have a user interface or is this a "headless process"?  Does this potentially process millions of results? What is the total count for?

**Comment by ulyChou** [comment]
I'm sorry. Maybe I didn't answer the question carefully enough
At present, I have a UI page for order management, which is a real scene. On this page, I need to realize many functions of multi-dimensional query for users, such as whether there are comments and whether there are multiple commodities. I cannot realize such requirements only through mysql. So I want to use ElasticSearch to solve my problem, and I want to tell the user how many orders with comments and how many orders with goods. This is similar to findCount, there are 100W orders per month, and the user needs to display 1000-2000 data at most on each page. But the user does need to have a pagination requirement, and I also need to show the user how many totalcounts the current query hit

**Comment by ulyChou** [comment]
![image](https://user-images.githubusercontent.com/46878819/140244191-0ce69914-e160-457f-81a0-162be6e13b54.png)

**Comment by ulyChou** [comment]
![image](https://user-images.githubusercontent.com/46878819/140244257-7d8f7a8d-a490-4ad0-8e58-aabc851a53da.png)

**Comment by ulyChou** [comment]
In 2019, we adopted Ebean as a formal ORM framework for production

**Comment by rbygrave** [comment]
> In 2019, we adopted Ebean as a formal ORM framework for production

That's great !

> use case requirements ...

- user needs to display 1000-2000 data at most on each page
- the user does need to have a pagination requirement
- show the user how many totalcounts the current query hit

Right, makes sense.  If I am correct the _"problem"_ is ...  

**_when the user clicks on "goto page 500" ... that is really inefficient_**

Now, assuming I have got that correct then, the solution to this is actually not let the user click "goto page 500". That is, what Google does and I think what most people do is that they only allow a user to page to some reasonable number like the first 10 pages.  They design the UX so that for a user to get to page 500 they have to _"goto page 10"_ and then _hit "next page" 490 times_ - the thinking is that not many users will actually do that, it is intentionally painful for users to do that, we use the UX to discourage users from doing the "goto page 500" query (and for good reason).

This design thinking is along the lines that in UX terms it is relatively easy to get to the first few pages, a user that wants to get to a really "deep" page instead refine the query more e.g. provide a more specific search criteria and execute another query.

If this is correct then we should be using `findPagedList()`.

Does that make sense?

**Comment by ulyChou** [comment]
> 他们设计了 UX，以便用户到达第 500 页时必须“转到第 10 页”然后点击“下一页”490 次
This is especially true for our scenario, where the user does not need to arbitrarily jump to a specified page.

But based on UX, we need to provide users with a pseudo-pagination scenario to meet their usage habits, and more importantly, the presence of totalCount, so that users know how many valid pieces of data their queries are getting.
So how do I use findEach in ebean-ElasticSearch to implement my use case

**Comment by rbygrave** [comment]
> how do I use findEach in ebean-ElasticSearch to implement my use case

We should not use `findEach()` for this use case.

We **should** be using `findPagedList()` and that returns PagedList which has the totalCount.

```java
    PagedList<Order> list = query.findPagedList();

   int totalCount = list.getTotalCount();
   List<Order>  orders = list.getList();
```
The previous recommendation on using `findEach()` was incorrect for this use case because I misunderstood the scenario and thought the use case was a query that processed lots of rows. For this use case we should indeed use `findPagedList()` and instead the "solution" to the problem of deep paging is to use UX design to prevent users from doing "deep paging".

**Comment by ulyChou** [comment]
All right so let's go back to the beginning of the problem.

```
query.setFirstRow(200).setMaxRows(200).setUseDocStore(true).findPagedList()
println(query.generatedSql)
```
```
{"from":200,"size":200,"query":{"bool":{"filter":{"range":{"createdAt":{"gte":"2021-10-04T11:50:47.447421300","lte":"2021-11-04T11:50:47.447421300"}}}}}}
```

I think for findPagedList we should add` "track_total_hits": true` to get the true totalCount

![image](https://user-images.githubusercontent.com/46878819/140255617-72d97c28-4137-4e90-b36f-4885ad2af1a8.png)
![image](https://user-images.githubusercontent.com/46878819/140255661-cad3225e-7d22-414e-8567-69aa149b317e.png)

**Comment by ulyChou** [comment]
And for these scenarios or use cases, I think it might be necessary for ebean-ElasticSearch to implement query.findcount () like mysql.
And if not, fix `"track_total_hits": True` After this problem, maybe I can use Ebean to get totalCount only by findPagedList()

**Comment by rbygrave** [comment]
Do you want to prepare a PR that adds track_total_hits ?

**Comment by ulyChou** [comment]
Ok, I will submit a PR  
`ebean-orm/ebean-elasticsearch`

**Comment by rbygrave** [comment]
Resolved with https://github.com/ebean-orm/ebean-elasticsearch/pull/43

Fixed in ebean-elastic 12.12.1
